### ISSUE TITLE
A command to remove indentations in multiline strings

### ISSUE THREAD

**Issue by kosciolek** [nushell/nushell#11477]
### Related problem

Embedded code is written in multiline strings.
```
def main [name: string] {
    $'
    .($name)-class-1 {
        color: red;
    }

    .($name)-class-2 {
        color: blue;
    }
    ' | save $"($name).scss"
}
```

Once it's saved to a file (or, say, piped into an interpreter), indents persist, making code less readable.

```

    .pkg2-class-1 {
        color: red;
    }

    .pkg2-class-2 {
        color: blue;
    }

    .pkg2-class-3 {
        color: green;
    }

    .pkg2-class-4 {
        color: aquamarine;
    }
    
```

Same goes for CSV and so on.

This is not good for scripts that write code to a file, which is frequent on a CI.

### Describe the solution you'd like

A way (possibly a `str unindent` subcommand) to remove the indentation, similar to Kotlin's [trimIndent](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.text/trim-indent.html).

### Describe alternatives you've considered

Do not indent the code. This makes code less readable IMO, especially if there's multi level nesting (all nu code's on the right, and embedded code is suddenly all to the left).
```
def main [name: string] {
$'
.($name)-class-1 {
    color: red;
}

.($name)-class-2 {
    color: blue;
}

.($name)-class-3 {
    color: green;
}

.($name)-class-4 {
    color: aquamarine;
}
' | save $"($name).scss"
}
```

**Comment by fdncred** [comment]
if you're generating nushell code, you'll eventually be able to use nufmt to format nushell code. people are working on that but it's not ready yet.

**Comment by NotTheDr01ds** [comment]
Yes, yes, yes - A million times yes.  I've considered opening a request for this, but I kind of feel like I've been asking too much (without being able to contribute enough in return) already.

I don't think that `nufmt` would work here unless this feature is implemented.  The embedded string in the Nushell code would still have a *different* indentation in code vs. in storage.

I hadn't seen `trimindent` yet, but this is very similar to the way that Nix [double single-quote strings work](https://nixos.org/manual/nix/stable/language/values).  I loved that design choice in Nix the moment I saw it.

**Comment by kosciolek** [comment]
I know a little bit of Rust, I'll be happy to implement it.

The algorithm seems simple and can be copied from Kotlin or Nix (I wonder what should happen when tabs are mixed with spaces, or if more exotic whitespace characters are used). I guess the main consideration is whether we'd like to implement it the Nix's way, or as a simple command like `str unindent`.

Could a maintainer comment on the spec?

**Comment by fdncred** [comment]
Seems like a `str unindent` would be a good way to start. If we decide that it should be part of strings in general, then we can just use that code to implement it.

**Comment by NotTheDr01ds** [comment]
@fdncred  Ah, I see what you're saying now -- You could format the string as desired (with base-indent) in source and then pass it to `str unindent` before use.  Yes, I think that would work as a first pass.

It at least seems to solve the readability problem with the source.

**Comment by NotTheDr01ds** [comment]
Nu implementation:

```Nushell
# Remove common indent from a multi-line string
export def "str unindent" []: string -> string {
  let text = $in
  let length = ($text | lines | length)

  let lines = $text | lines

  # Determine if first and/or last lines are empty and should be dropped
  let doNothing = {||}
  let ifSkipFirst = match ($lines | first | str trim) {
    "" => {{skip}}
    _ => {$doNothing}
  }
  let ifDropLast = match ($lines | last | str trim) {
    "" => {{drop}}
    _ => {$doNothing}
  }

  # Execute the conditional skip/drop logic
  let lines = (
    $lines
    | do $ifSkipFirst
    | do $ifDropLast
  )
  | # Convert list to table
  | wrap text

  let minimumIndent = (
    # Add a column to each row with the number of leading spaces
    $lines | insert indent {|line|
      if ($line.text | str trim | is-empty) {
        # If the line contains only whitespace, don't consider it
        null
      } else {
        $line.text
        | parse -r '^(?<indent> +)'
        | get indent.0?
        | default ''
        | str length
      }
    }
    | # And return the minimum
    | get indent
    | math min
  )

  let spaces = ('' | fill -c ' ' -w $minimumIndent)

  $lines
  | update text {|line|
      $line.text 
      | str replace -r $'^($spaces)' ''
    }
  | get text
  | to text

}
```

**Comment by NotTheDr01ds** [comment]
Notes:

* If the first or last lines consist only of whitespace, they are discarded.
* Other lines consisting only of whitespace are disregarded when calculating indentation level. Otherwise, the string in the original example would have 0 spaces removed due to the blank line before `.($name)-class-2 {`
* Tabs *`\t`) are not considered at the moment for indentation purposes; only spaces.

Usage:

```Nushell
def main [name: string] {
    $'
    .($name)-class-1 {
        color: red;
    }

    .($name)-class-2 {
        color: blue;
    }
    ' | str unindent | save $"($name).scss"
}
```

**Comment by NotTheDr01ds** [comment]
@fdncred  Thoughts on whether this might be appropriate for `std`?

**Comment by jameschensmith** [comment]
I don't know if this helps at all, but I came up with another approach to removing indentations. My approach tries to follow [Swift's multiline string literals](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/stringsandcharacters#Multiline-String-Literals). Essentially, indentation comes from last line of string, and first & last lines are trimmed from final string.

```nushell
export def "str unindent" []: string -> string {
    let text = $in
    if not ($text | str starts-with (char newline)) {
        return $text
    }

    let lines = $text
    | split row (char newline)

    let indent = $lines
    | last
    | split chars
    | if ($in | all {|c| $c == (char space)}) {
        length
    } else {
        error make {
            msg: 'last line must contain only whitespace'
        }
    }

    $lines
    | range 1..-2
    | each {|line|
        $line
        | str substring ..$indent
        | split chars
        | if ($in | all {|c| $c == (char space)}) {
            $line | str substring $indent..
        } else {
            error make {
                msg: 'insufficient indentation of line'
            }
        }
    }
    | str join (char newline)
}
```

**Comment by NotTheDr01ds** [comment]
@jameschensmith The Swift implementation sounds interesting.  In theory, it's a simpler implementation, and it should usually yield the same results.  But there seems to be a slight issue in that implementation that I ran across by accident on my first try.  Whitespace in the first line, immediately after the quote, breaks the function entirely - the string is returned as originally indented.  Reading the Swift spec, I don't think that should be the case, right?

Best demonstrated if you use the GitHub "Copy" icon and then paste it directly into the shell:

```
"  
  one
  two
  three
  " | str unindent
```

... returns a string where all lines are still indented by 2 spaces. It doesn't matter how many "hidden" spaces on the first line are there.

**Comment by NotTheDr01ds** [comment]
Also a question - In Swift, what's the behavior when there is a blank line with no whitespace?  (Again, using "Copy" to test the whitespace properly):

```Nushell
"
  One

  Two

  Three
  "  | str unindent
```

It seems to me that the string should be unindented, with the blank line's indentation ignored, which is how I implemented it.  In your case, you would error out though - Does that match the Swift behavior?

**Comment by jameschensmith** [comment]
Regarding first line. I admit my current implementation handles it a bit too simply. Swift will actually error if there is anything on the first line (whitespace or otherwise). I will update my command to be equivalent.

Regarding blank lines. Mine does not error on the example you provided. The way Swift handles this, according to the documentation, is to keep the line break. I say according to the documentation because if you play around with [SwiftFiddle](https://swiftfiddle.com/), it appears to just remove the line break whether there is no indent or whether the indent matches the last line. It only keeps the line break if the indent is greater than the last line. My implementation keeps the line breaks. It could possibly be made to handle more scenarios. I would hope that maybe someday, Nushell could natively support ignoring indentation in string literals.

**Comment by rosshadden** [comment]
Any unindent command we add (big 👍!) should handle tabs too. I know it was mentioned that the above attempts do not do so, but I think a standard version we add should take them into account. I might try making a pass at this if I get time this week.

**Comment by fdncred** [comment]
I have a plugin that has [str dedent](https://github.com/fdncred/nu_plugin_strutils/blob/main/src/commands/str_dedent.rs) in it that uses the [textwrap crate](https://docs.rs/textwrap/0.16.2/textwrap/fn.dedent.html). I believe it uses is_whitespace() to figure things out.

**Comment by in-pursuit-of-ash** [comment]
~~`std-rfc/str dedent` is great, but it'd be nice for it to also accept empty lines, as some editors (reasonably, IMO) strip trailing whitespace, therefore breaking the command, which, currently, expects every line to be indented~~

EDIT: Nevermind; it works as I'd expect. I just messed up while testing. Sorry for the unnecessary notification. 😅
